voxblox/include/voxblox/skeletons/neighbor_tools.h

```diff
@@ -0,0 +1,177 @@
+  for (unsigned int i = 0; i < 3; ++i) {
+    for (int j = -1; j <= 1; j += 2) {
+      direction(i) = j;
+      getNeighbor(block_index, voxel_index, direction, &neighbor.first,
+                  &neighbor.second);
+      neighbors->emplace_back(neighbor);
+      distances->emplace_back(1.0);
+      directions->emplace_back(direction);
+    }
+    direction(i) = 0;
+  }
+  if (connectivity > 6) {
+    // Distance sqrt(2) set.
+    for (unsigned int i = 0; i < 3; ++i) {
+      unsigned int next_i = (i + 1) % 3;
+      for (int j = -1; j <= 1; j += 2) {
+        direction(i) = j;
+        for (int k = -1; k <= 1; k += 2) {
+          direction(next_i) = k;
+          getNeighbor(block_index, voxel_index, direction, &neighbor.first,
+                      &neighbor.second);
+          neighbors->emplace_back(neighbor);
+          distances->emplace_back(kSqrt2);
+          directions->emplace_back(direction);
+        }
+        direction(i) = 0;
+        direction(next_i) = 0;
+      }
+    }
+  }
+
+  if (connectivity > 18) {
+    // Distance sqrt(3) set.
+    for (int i = -1; i <= 1; i += 2) {
+      direction(0) = i;
+      for (int j = -1; j <= 1; j += 2) {
+        direction(1) = j;
+        for (int k = -1; k <= 1; k += 2) {
+          direction(2) = k;
+          getNeighbor(block_index, voxel_index, direction, &neighbor.first,
+                      &neighbor.second);
+          neighbors->emplace_back(neighbor);
+          distances->emplace_back(kSqrt3);
+          directions->emplace_back(direction);
+        }
+      }
+    }
+  }
+}
+
+template <typename VoxelType>
+void NeighborTools<VoxelType>::getNeighbor(
+    const BlockIndex& block_index, const VoxelIndex& voxel_index,
+    const Eigen::Vector3i& direction, BlockIndex* neighbor_block_index,
+    VoxelIndex* neighbor_voxel_index) const {
+  CHECK_NOTNULL(layer_);
+  DCHECK(neighbor_block_index != NULL);
+  DCHECK(neighbor_voxel_index != NULL);
+
+  *neighbor_block_index = block_index;
+  *neighbor_voxel_index = voxel_index + direction;
+
+  for (unsigned int i = 0; i < 3; ++i) {
+    while ((*neighbor_voxel_index)(i) < 0) {
+      (*neighbor_block_index)(i)--;
+      (*neighbor_voxel_index)(i) += voxels_per_side_;
+    }
+    while ((*neighbor_voxel_index)(i) >=
+           static_cast<IndexElement>(voxels_per_side_)) {
+      (*neighbor_block_index)(i)++;
+      (*neighbor_voxel_index)(i) -= voxels_per_side_;
+    }
+  }
+}
+
+template <typename VoxelType>
+Eigen::Vector3i NeighborTools<VoxelType>::getOffsetBetweenVoxels(
+    const BlockIndex& start_block_index, const VoxelIndex& start_voxel_index,
+    const BlockIndex& end_block_index,
+    const VoxelIndex& end_voxel_index) const {
+  BlockIndex current_block_index = end_block_index;
+  Eigen::Vector3i voxel_offset = Eigen::Vector3i::Zero();
+
+  // Line up the blocks so that they're in the same block.
+  for (unsigned int i = 0; i < 3; ++i) {
+    while (start_block_index(i) > current_block_index(i)) {
```

Comment:
Am I missing something why does this need to be a loop, can't we just get everything from turning the local indexes into global indexes and then taking the difference?